src/MainWindow.vala

```diff
@@ -570,35 +573,34 @@ namespace PantheonTerminal {
         }
 
         private void restore_saved_state (bool restore_pos = true) {
-            saved_tabs = saved_state.tabs;
-            default_width = PantheonTerminal.saved_state.window_width;
-            default_height = PantheonTerminal.saved_state.window_height;
+            saved_tabs = PantheonTerminal.TerminalApp.saved_state.get_strv ("tabs");
 
-            Gdk.Rectangle geometry;
-            get_screen ().get_monitor_geometry (get_screen ().get_primary_monitor (), out geometry);
+            var rect = Gdk.Rectangle ();
+            PantheonTerminal.TerminalApp.saved_state.get ("window-size", "(ii)", out rect.width, out rect.height);
 
-            if (default_width == -1) {
-                default_width = geometry.width * 2 / 3;
-            }
+            default_width = rect.width;
+            default_height = rect.height;
+
+            if (default_width == -1 || default_height == -1) {
+                Gdk.Rectangle geometry;
+                get_screen ().get_monitor_geometry (get_screen ().get_primary_monitor (), out geometry);
 
-            if (default_height == -1) {
+                default_width = geometry.width * 2 / 3;
                 default_height = geometry.height * 3 / 4;
             }
 
             if (restore_pos) {
-                int x = saved_state.opening_x;
-                int y = saved_state.opening_y;
+                int window_x, window_y;
+                PantheonTerminal.TerminalApp.saved_state.get ("window-position", "(ii)", out window_x, out window_y);
```

Comment:
You could use the already available `rect` instead of defining new window x and y.
```suggestion
                PantheonTerminal.TerminalApp.saved_state.get ("window-position", "(ii)", out rect.x, out rect.y);
```